Added Kafka clients instrumentation blog post

ppatierno opened the pull request:
This PR addresses #1668 about writing a blog post on Kafka clients instrumentation with OpenTelemetry.

svrnm commented on content/en/blog/2022/instrument-kafka-clients/index.md:
```diff
@@ -0,0 +1,252 @@
+author: "[Paolo Patierno](https://github.com/ppatierno)"
+---
+
+Nowadays, [Apache Kafka](https://kafka.apache.org/) is chosen as the central piece in a distributed system.
```
> the central piece in a distributed system.

the central piece for what? I as a reader have the feeling there is something missing or should it be "a central piece"?

ppatierno replied:
The idea behind this was to say that it's used a the "nervous" system for communication between applications in a distributed system.

svrnm replied:
I would be ok with using that term "nervous system", I guess most people understand that picture.

svrnm commented on content/en/blog/2022/instrument-kafka-clients/index.md:
```diff
@@ -0,0 +1,252 @@
+---
+
+Nowadays, [Apache Kafka](https://kafka.apache.org/) is chosen as the central piece in a distributed system.
+Different services communicate each other by using Apache Kafka as a messaging system but even more as en event or data streaming platform.
```
communicate **with** each other.

svrnm commented on content/en/blog/2022/instrument-kafka-clients/index.md:
```diff
@@ -0,0 +1,252 @@
+It takes care of the overall Kafka infrastructure, while you can focus on developing applications which use it.
+
+Within the overall picture, because of the distributed nature, it is quite difficult to track how messages are moved around.
+This is where the OpenTelemetry project comes into the picture.
```
the OpenTelemetry project -> OpenTelemetry

svrnm commented on content/en/blog/2022/instrument-kafka-clients/index.md:
```diff
@@ -0,0 +1,252 @@
+
+Within the overall picture, because of the distributed nature, it is quite difficult to track how messages are moved around.
+This is where the OpenTelemetry project comes into the picture.
+It defines the specification of semantic conventions for [messaging systems](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md) as well as it provides multiple instrumentation libraries for adding tracing to messaging based applications.
```
Although this it the otel blog, the reader might not know what semantic conventions are. Maybe skip that for now and rephrase that to:

This is where OpenTelemetry comes into the picture: it provides multiple instrumentation libraries for adding tracing to messaging based applications. Of course, there is one for Apache Kafka clients.

ppatierno replied:
I thought it was useful to mention the messaging semantic conventions because I am referring to them in the `Instrumentation in action` section as well [here](https://github.com/open-telemetry/opentelemetry.io/pull/1677/files#diff-1713a18cb9f06d7587ef469a52bb57100157ee894ee5a3aa6bf40ecd46147797R277), to explain that current "output" in terms of trace, spans and relationship between them is evolving and going to change with the OTEP.

svrnm replied:
then put the instrumentation first and the semantic conventions second? People are more interested in the first I would say.

svrnm commented on content/en/blog/2022/instrument-kafka-clients/index.md:
```diff
@@ -0,0 +1,252 @@
+It defines the specification of semantic conventions for [messaging systems](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md) as well as it provides multiple instrumentation libraries for adding tracing to messaging based applications.
+Of course, there is one for Apache Kafka clients.
+
+But usually, the architecture can even be more complicated, having applications not able to connect directly to the Apache Kafka cluster and talking its own custom protocol but using a different one, like for example HTTP.
```
But usually, the architecture can even be more complicated, to But usually, the architecture can even be more complicated:

I have the feeling that the `:` makes it clearer to the reader that what is following is an explanation of the things said in the sentence before.

svrnm commented on content/en/blog/2022/instrument-kafka-clients/index.md:
```diff
@@ -0,0 +1,252 @@
+Of course, there is one for Apache Kafka clients.
+
+But usually, the architecture can even be more complicated, having applications not able to connect directly to the Apache Kafka cluster and talking its own custom protocol but using a different one, like for example HTTP.
+In this case, tracing how messages are produced and consumed via HTTP through Apache Kafka could be really complex.
```
could be -> is

svrnm commented on content/en/blog/2022/instrument-kafka-clients/index.md:
```diff
@@ -0,0 +1,252 @@
+
+But usually, the architecture can even be more complicated, having applications not able to connect directly to the Apache Kafka cluster and talking its own custom protocol but using a different one, like for example HTTP.
+In this case, tracing how messages are produced and consumed via HTTP through Apache Kafka could be really complex.
+Even the Strimzi project provides a bridge, with the OpenTelemetry support, for adding tracing data by using the corresponding instrumentation library.
```
Why 'Even'? I think you can skip that.

svrnm commented on content/en/blog/2022/instrument-kafka-clients/index.md:
```diff
@@ -0,0 +1,252 @@
+Different services communicate each other by using Apache Kafka as a messaging system but even more as en event or data streaming platform.
+
+Taking into account the cloud-native approach for developing microservices, quite often [Kubernetes](https://kubernetes.io/) is also used to run the workloads.
+In this scenario, you can also easily deploy and manage an Apache Kafka cluster on top of it, by using a project like [Strimzi](https://strimzi.io/).
```
strimzi is another cncf project, right?

ppatierno replied:
Yes it is. What's the point you would like to make? Maybe that there is an "integration" between Strimzi and OpenTelemetry as part of the CNCF? Adding something like this to the blog post?

svrnm replied:
first of all I wanted to verify that it's another OSS solution and not something commercial.

we might consider cross-posting this blog post in a second step on the cncf blog?

ppatierno replied:
We can for sure cross-posting from our Strimzi official web site. We have never published a post on the CNCF blog.

svrnm commented on content/en/blog/2022/instrument-kafka-clients/index.md:
```diff
@@ -0,0 +1,252 @@
+In this case, tracing how messages are produced and consumed via HTTP through Apache Kafka could be really complex.
+Even the Strimzi project provides a bridge, with the OpenTelemetry support, for adding tracing data by using the corresponding instrumentation library.
+
+In this article, we are going to see how it is possible to enable tracing on Apache Kafka based client applications in different ways.
```
we are going to see -> you will learn.

svrnm commented on content/en/blog/2022/instrument-kafka-clients/index.md:
```diff
@@ -0,0 +1,252 @@
+
+The main environment variables to be set are the following:
+
+* `OTEL_SERVICE_NAME`: specify the logical service name. This is really useful when using a tracing UI (i.e. Jaeger UI) for showing data.
```
> This is really useful when using a tracing UI (i.e. Jaeger UI) for showing data.

The service name is kind of mandatory, maybe you can phrase this even stronger?

ppatierno replied:
Actually in one of the previous discussion with other committers (I don't remember where and who :-D) they said that I should not make the point that service name is mandatory, which is kind of true, but at same time I agree with you. In general, it doesn't make sense not having a service name. For example, you could skip setting it but then in the Jaeger UI, for example, you have "unknown service" in the drop down list for selecting the service for getting the traces.

svrnm replied:
yeah, "kind of mandatory", maybe write something like "It is recommended to set it"?

svrnm commented on content/en/blog/2022/instrument-kafka-clients/index.md:
````diff
@@ -0,0 +1,252 @@
+
+```java
+Resource resource = Resource.getDefault()
+        .merge(Resource.create(Attributes.of(ResourceAttributes.SERVICE_NAME, "my-kafka-service")));
````
above you ask people to use the environment variable / system property, now you're setting the SERVICE_NAME via code as well, that might confuse the user.

ppatierno replied:
This is the section related to not using the SDK auto-configuration (which relies on env vars as well) but it's about creating the `OpenTelemetry` instance programmatically from your own code.

svrnm replied:
ah, gotcha

svrnm commented on content/en/blog/2022/instrument-kafka-clients/index.md:
```diff
@@ -0,0 +1,252 @@
+
+Then use the `tracingConsumer` as usual for receiving messages from the Kafka cluster.
+
+## Instrumenting by using the agent
```
I personally would like seeing that automatic way first, people (including myself) are lazy, so having something they can do within a few "clicks" is always a better hook than copying a bunch of code snippets.

ppatierno replied:
That's a good point. I will re-arrange the order!

ppatierno replied:
Done!

svrnm commented on content/en/blog/2022/instrument-kafka-clients/index.md:
```diff
@@ -0,0 +1,252 @@
+## Instrumentation in action
+
+In order to practice with the instrumentation of Kafka clients by using the provided examples, first of all you need an Apache Kafka cluster.
+The easiest way is to download it from the [official website](https://kafka.apache.org/downloads) and running just one ZooKeeper node together with one Kafka broker.
```
isn't there some docker images people can use, or even a docker compose that build everything for them?

ppatierno replied:
There are ... some of them use Kafka images from the Strimzi project, others use the Confluent one. I would like to be less opinionated and using directly the upstream project.

svrnm replied:
What I wanted to say: is there one in your sample github repo or in one you create for this blog post, people can just copy that and run it (like we did it in the nginx blog where we give something people can follow step-by-step from 0)

ppatierno replied:
No, because most of my development with Kafka is always with running it locally without Docker compose (I don't have it installed at all :-)). The other is about using Kubernetes when I worked on Strimzi but using Kubernetes is really useless and too much for the purpose of the blog post.

svrnm replied:
ok, then let's leave it like that!

svrnm reviewed (changes requested):
Thanks for writing this up! I provided some feedback via my review.

Please also run `npx prettier --write content/en/blog/2022/instrument-kafka-clients/index.md` to get some automatic style updates (line width, etc.)

ppatierno commented:
@svrnm I ran the "prettier" command but imho it makes review harder. As our experience about writing on the Strimzi blog, we prefer having "one sentence per line" because you can comment on each sentence directly for providing feedback.
Maybe my mistake was to run the "prettier" too soon after your request. I should have run it at the end, when the PR was finally approved.

svrnm commented:
> @svrnm I ran the "prettier" command but imho it makes review harder. As our experience about writing on the Strimzi blog, we prefer having "one sentence per line" because you can comment on each sentence directly for providing feedback. Maybe my mistake was to run the "prettier" too soon after your request. I should have run it at the end, when the PR was finally approved.

Makes sense, then let's leave it like that and we run the prettier before merging.

ppatierno commented:
@svrnm I think I addressed your comments, is there anything more? Any other maintainers to review this? At this point I should also change the publication date (which is tomorrow Sep 5th).

svrnm commented on content/en/blog/2022/instrument-kafka-clients/index.md:
```diff
@@ -0,0 +1,320 @@
+---
+title: Instrumenting Apache Kafka clients with OpenTelemetry
+linkTitle: Instrumenting Apache Kafka clients with OpenTelemetry
```
Replace with "Instrumenting Apache Kafka clients" to make the short title even shorter, we did the same for many other blog posts, see on the left side menu here:

https://deploy-preview-1677--opentelemetry.netlify.app/blog/2022/instrument-kafka-clients/

ppatierno replied:
Makes sense. Done!

svrnm commented:
@ppatierno minor thing, see above, put nothing that stops us from mergin.

@open-telemetry/blog-approvers: please review & approve, thanks!

austinlparker reviewed (approved):
Lgtm
